### PR TITLE
easipro

### PR DESCRIPTION
**Pull request by bradymiller**
Ported this code from 5.0.0: https://github.com/bradymiller/openemr/commits/easipro-base-5-0-0

Can see details on easipro here: https://sites.northwestern.edu/easipro/

Issues working on:
- ~~Update code to 5.0.2 standards (plan to migrate core functionalities to a Easipro class; make a interface/easipro directory to hold the main gui script along with a backend script for ajax calls)~~ (DONE)
- ~~Relabel assessments sql table to something more specific (pro_assessments)~~ (DONE)
- ~~Bring in patient portal stuff from the 5.0.0 code into the new patient portal~~ (DONE)
- ~~Fix expiration date stuff (and ensure other 2 dates are populated correctly)~~ (DONE)
- ~~For production, will need a production server to connect to (now have credentials to use the server for development purposes) and licensing issue sorted out~~ (DONE)

For licensing options for this feature, contact api@assessmentcenter.net

Can get developer sandbox credentials for server https://www.assessmentcenter.net/ac_api by getting an account at https://www.assessmentcenter.net

### CODE REVIEW

**Comment by bradymiller** [comment]
This feature is now working using a development connection (hard-coded in the Easipro class for now) to the https://www.assessmentcenter.net/ server. Now need to figure out licensing/mechanism for OpenEMR users.

To see how it works in this demo, check out:
1. Ensure it is turned on at Administration->Globals->Connectors->Enable Easipro

2. Request the PRO assessments/tests from patient summary screen at PRO link
image
image
image
image
image

3. Patient can then perform these ordered tests on the patient portal:
image
image
image

**Comment by sjpadgett** [comment]
> This feature is now working using a development connection

I'll can add support for this in my PRO upcoming offering because this vendor won't be the only source for questionnaires and scoring.. I didn't know this was even in v5.0.0 but nicely done Brady. Say again, very nicely done and thanks. um seems like you've done these kind of thing before. Who knew.:)
Really, very happy to see this.......

**Comment by bradymiller** [comment]
To clarify, this wasn't in 5.0.0. It was from work that a developer, @StrongTSQ , did based on 5.0.0 code. I basically just brought it to master standards along with bunch of refactoring. This PRO stuff is really cool; trying to figure out a way for licensing costs to be more manageable.

**Comment by sjpadgett** [comment]
> trying to figure out a way for licensing costs to be more manageable

I'll try as well.

File: src/Easipro/Easipro.php [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+use MyMailer;
+use OpenEMR\Common\Http\oeHttp;
+
+class Easipro
+{
+    public function __construct()
+    {
+        // direct to easipro dev server
+        $this->server = "https://www.assessmentcenter.net/ac_api/2014-01/";
```
Comment (bradymiller) [inline src/Easipro/Easipro.php]: 2014-01 migrate to call

File: src/Easipro/Easipro.php [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+        // credential token for easipro dev server
+        $this->devAuth = "QkJENjI5MzUtRjc2Ri00RUM4LTg4MzQtQkRBQTc1REFEOEFCOjlBMzVEMzEzLUU3QkMtNDFDOS04OTMzLTNBM0Q3Mzk1M0Y3Mw==";
```
Comment (bradymiller) [inline src/Easipro/Easipro.php]: decode and submit as user/pass

File: portal/get_pro.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        <th><?php echo xlt('Deadline (CST)'); ?></th>
+        <th><?php echo xlt('Status'); ?></th>
+        <th>&nbsp;</th>
+    </tr>
+    <?php
+    foreach ($records as $row) {
+        echo "<tr>";
+        echo "<td>".text($row['form_name'])."</td>";
+        echo "<td>".text($row['deadline'])."</td>";
+        echo "<td id='asst_status_" . text($row['assessment_oid']) . "'>".text($row['status'])."</td>";
```
Comment (bradymiller) [inline portal/get_pro.php]: attr()

File: portal/get_pro.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        if ($row['status']=='ordered') {
+            echo "<td id='asst_" . text($row['assessment_oid']) . "'><button class='btn btn-sm btn-primary' onclick=\"startAssessment(this," . attr_js($row['assessment_oid']) . ")\">" . xlt('Start Assessment') . "</button></td>";
```
Comment (bradymiller) [inline portal/get_pro.php]: attr()

**Comment by bradymiller** [comment]
Testing well and bringing this feature into codebase.
Thank you @StrongTSQ !!!!

**Comment by StrongTSQ** [comment]
Pleasure to see that going to production!
-- 


Shiqiang Tao, PhD

Research Assistant Professor

Department of Neurology

McGovern Medical School

UT Health Science Center at Houston

Phone: 713-500-7720

**Pull request merged**